qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -12,117 +12,479 @@
+    to x, y, and z.  2nd axis represents the index of terms.
+    3rd axis represents the index of register.
+    `coeffs` is one-dimensional `ndarray` with the length of the number of terms.
+
+    **Algebra**
+
+    :class:`SpinOp` supports the following basic arithmetic operations: addition, subtraction,
+    scalar multiplication, and dagger(adjoint).
+    For example,
+
+    .. jupyter-execute::
+
+        from qiskit_nature.operators import SpinOp
+
+        x = SpinOp("X", spin=3/2)
+        y = SpinOp("Y", spin=3/2)
+        z = SpinOp("Z", spin=3/2)
+
+        print("Raising operator:")
+        print(x + 1j * y)
+        plus = SpinOp("+", spin=3/2)
+        print("This is same with: ", plus)
+        print("Lowering operator:")
+        print(x - 1j * y)
+        minus = SpinOp("-", spin=3/2)
+        print("This is same with: ", minus)
+
+        print("Dagger")
+        print(~(1j * z))
+
     """
-    Spin type operators. This class represents sums of `spin strings`, i.e. linear combinations of
-    SpinOperators with same spin and register length.
-    """
 
-    def __init__(self, operator_list):
-        # 1. Parse input
-        self._register_length = len(operator_list[0])
-        self.spin = operator_list[0].spin
-        for elem in operator_list:
-            assert isinstance(elem, SpinOperator)
-            assert len(elem) == self._register_length, \
-                'Cannot sum operators acting on registers of different length'
-            assert elem.spin == self.spin, \
-                'Cannot sum operators with different spins.'
+    _XYZ_DICT = {"X": 0, "Y": 1, "Z": 2}
+    _VALID_LABEL_PATTERN = re.compile(
+        r"^([IXYZ\+\-]_\d(\^\d)?\s)*[IXYZ\+\-]_\d(\^\d)?(?!\s)$|^[IXYZ\+\-]+$"
+    )
+    _SPARSE_LABEL_PATTERN = re.compile(r"^([IXYZ]_\d(\^\d)?\s)*[IXYZ]_\d(\^\d)?(?!\s)$")
+    _DENSE_LABEL_PATTERN = re.compile(r"^[IXYZ]+$")
+
+    def __init__(
+        self,
+        data: Union[
+            str,
+            List[Tuple[str, complex]],
+            Tuple[np.ndarray, np.ndarray],
+        ],
+        spin: Union[float, Fraction] = Fraction(1, 2),
+    ):
+        r"""``SpinOp``.
+
+        Args:
+            data: label string or list of labels and coefficients. See documentation of
+                  :class:`SpinOp` for more details.
+            spin: positive half-integer (integer or half-odd-integer) that represents spin.
+
+        Raises:
+            ValueError: invalid data is given.
+            QiskitNatureError: invalid spin value
+        """
+        self._coeffs: np.ndarray
+        self._spin_array: np.ndarray
+        dtype = np.complex128  # TODO: configurable data type. mixin?
+
+        if (round(2 * spin) != 2 * spin) or (spin <= 0):
+            raise QiskitNatureError(
+                f"spin must be a positive half-integer (integer or half-odd-integer), not {spin}."
+            )
+        self._dim = int(round(2 * spin)) + 1
```

Comment:
Since we decided to go with `Fraction` we should actually leverage that and get rid of `round`. I propose something along the lines of the following:
```suggestion
        spin = Fraction(spin)
        
        if (spin.denominator not in (0, 1)) or (spin <= 0):
            raise QiskitNatureError(
                f"spin must be a positive, either integral or half-integral value, not {spin}."
            )
        self._dim = int(2 * spin + 1)
```